### PR TITLE
Feature Recursive TV

### PR DESCRIPTION
**Pull request by JingkangZhang**
@abizer @fydai @cg505 
Screen Shot 2019-09-05 at 11 51 29 AM
A recursive TV screen synchronous with the main screen. Only nested one level deep.

### CODE REVIEW

**Comment by ja5087** [comment]
Minor: can you remove .DS_Store and also add it to the gitignore?

**Comment by MCer4294967296** [comment]
I'd say this is an interesting feature that our users might not realize for a while.

btw his ocf username is zjk; he is the tall guy with long hair and is with me or jerryzhou a lot of the time.

**Review by abizer — APPROVED** [review]
lmao nice, pending @ja5087's comments

File: www/index.html [hunk 1]
```diff
@@ -425,170 +426,1535 @@
...
-                </g>
-                <g id="g3817_3_">
-                    <g id="g3198_3_">
-                        <path id="path3200_3_" class="st6" d="M1356.7,230.4c-0.8-0.5-0.5,1.3-0.8,2.6c-0.2,0.7,2.2,1.9,2.9,1.1
-                            C1358.9,233.9,1356.7,230.4,1356.7,230.4"/>
-                    </g>
-                    <g id="g3202_3_">
-                        <path id="path3204_3_" class="st6" d="M1355.9,235.5c0,2.1-0.5,3.1-1.8,4.5c-1.2,1.4-2.4,1.9-3.1,2.5c-0.6,0.6-0.5,1,0.4,1.6
-                            c0.9,0.6,1.7,1.4,2.6,0.8c0.8-0.6,2.9-2.4,3.7-4.1c0.8-1.6,1-2.6,1.3-3.8c0.2-1.2-3.6-1.9-3.6-1.9"/>
+    <!-- To obtain a recursive TV screen effect, below svg was copied from the entire original svg with serval changes.
```
Comment (jvperrin) [inline www/index.html]: nit: `serval` -> `several`

**Review by jvperrin — COMMENTED** [review]
Nice, this is pretty fun!

Is there any way to avoid duplicating the whole SVG with a few changes? Could it be put into a separate file and included twice or something similar?

**Comment by JingkangZhang** [comment]
> Nice, this is pretty fun!
> 
> Is there any way to avoid duplicating the whole SVG with a few changes? Could it be put into a separate file and included twice or something similar?

@jvperrin 
While this might indeed be able to reduce the file size by around 100kb, the current version is more explicit and maintainable in my opinion, for otherwise to make the changes in code (rather than by hand) would require some SVG parsing and there are changes to more than 30 locations between the two SVGs. Also loading speed might not be a major concern for this page.

File: www/index.html (hunk 1 above)
Comment (JingkangZhang) [reply www/index.html]: fixed

File: package.json [hunk 2]
```diff
@@ -1,25 +1,25 @@
...
+    "devDependencies": {
+        "sass": "^1.22.10",
+        "tslint": "^5.19.0",
+        "typescript": "^3.6.2"
+    },
+    "scripts": {
+        "tsc": "tsc",
+        "tsc:w": "tsc -w",
+        "tslint": "tslint -p tsconfig.json"
+    }
```
Comment (BernardZhao) [inline package.json]: Why is this JSON file being indented with 4 spaces now?

**Review by BernardZhao — COMMENTED** [review]
I would recommend separating out the formatting changes so that we can actually properly review the relevant changes.

Since I don't know what was explicitly changed, the only tip I can give is that be careful about reusing gradients if you did so, since they share CSS ids, and classes affecting one will affect the other. I guess since this is supposed to be a reflection of the current screen though, so in reality I guess that would be the actual intended behavior 💯.

File: package.json (hunk 2 above)
Comment (JingkangZhang) [reply package.json]: Oops that was my vscode's prettify on save.

**Comment by JingkangZhang** [comment]
> I would recommend separating out the formatting changes so that we can actually properly review the relevant changes.
> 
> Since I don't know what was explicitly changed, the only tip I can give is that be careful about reusing gradients if you did so, since they share CSS ids, and classes affecting one will affect the other. I guess since this is supposed to be a reflection of the current screen though, so in reality I guess that would be the actual intended behavior 💯.

@BernardZhao 
Sharing class names was indeed the intended behavior. What I did have to fix while copying was to substitute some ids with classNames and removing some ids for the inner svg.

**Comment by JingkangZhang** [comment]
Wait I just found that www/script.js was not tracked and not pushed to my fork.
Screen Shot 2019-09-05 at 2 53 07 PM
Why ignored it? Wait, where did I pull this file from earlier if it's not in there on master?

**Comment by JingkangZhang** [comment]
> Wait I just found that www/script.js was not tracked and not pushed to my fork.
> Screen Shot 2019-09-05 at 2 53 07 PM
> Why ignored it? Wait, where did I pull this file from earlier if it's not in there on master?

Okay, so at one point, the project changed to using TypeScript. I'll rewrite my code to script.ts to conform to the change.

**Comment by jvperrin** [comment]
> > Nice, this is pretty fun!
> > Is there any way to avoid duplicating the whole SVG with a few changes? Could it be put into a separate file and included twice or something similar?
> 
> @jvperrin
> While this might indeed be able to reduce the file size by around 100kb, the current version is more explicit and maintainable in my opinion, for otherwise to make the changes in code (rather than by hand) would require some SVG parsing and there are changes to more than 30 locations between the two SVGs. Also loading speed might not be a major concern for this page.

Yeah I'm not concerned about file size or loading speed here really since it's not loaded by anyone else and doesn't matter if it takes a few seconds or whatever to render on the TV, so I think we're both focused on maintainability concerns. I'm still a bit concerned that the SVGs are large blobs of mostly unreadable code (at least not readable by hand), and that having them in a separate file would be nice, but maybe the duplication between them can't actually be reduced. Are there any parts of each SVG that are the same and could be factored out into a common sub-svg file?

**Comment by JingkangZhang** [comment]
> > > Nice, this is pretty fun!
> > > Is there any way to avoid duplicating the whole SVG with a few changes? Could it be put into a separate file and included twice or something similar?
> > 
> > 
> > @jvperrin
> > While this might indeed be able to reduce the file size by around 100kb, the current version is more explicit and maintainable in my opinion, for otherwise to make the changes in code (rather than by hand) would require some SVG parsing and there are changes to more than 30 locations between the two SVGs. Also loading speed might not be a major concern for this page.
> 
> Yeah I'm not concerned about file size or loading speed here really since it's not loaded by anyone else and doesn't matter if it takes a few seconds or whatever to render on the TV, so I think we're both focused on maintainability concerns. I'm still a bit concerned that the SVGs are large blobs of mostly unreadable code (at least not readable by hand), and that having them in a separate file would be nice, but maybe the duplication between them can't actually be reduced. Are there any parts of each SVG that are the same and could be factored out into a common sub-svg file?

I'm yet to find out a non-hacky way to modify a base SVG to the two desired versions (inner and main), so I currently hard-code it. As for maintainability, with browser dev tools, future contributors will easily be able to identify the parts that they want to change. Whereas if I use hacky placeholdering in a base SVG, and later matching and inserting to fork the base SVG into two desired SVGs, future changes would also have to be carried out this way rather than directly modifying the nest SVG code.

File: package.json (hunk 2 above)
Comment (JingkangZhang) [reply package.json]: fixed

**Comment by kkuehlz** [comment]
Hmm. I love the feature, although I do share some of @jvperrin's concerns. The CSS classes are good for cosmetic changes, although minor changes to svg itself now require code to change in two places, which I really don't like. Would a lightweight templating engine that supports partials be less hacky?

**Comment by JingkangZhang** [comment]
> Hmm. I love the feature, although I do share some of @jvperrin's concerns. The CSS classes are good for cosmetic changes, although minor changes to svg itself now require code to change in two places, which I really don't like. Would a lightweight templating engine that supports partials be less hacky?

May I look into Vue for this?

**Comment by BernardZhao** [comment]
> May I look into Vue for this?

It's funny that you mention this, but I also did something similar when working with SVGs on my experimental OCF project [here](https://github.com/BernardZhao/ocf/blob/master/src/components/Logo.vue). I used Vue to wrap an svg element, so this might be a good example. You can also see the problem that I mentioned earlier, being that component reuse also messes up id tagging when reused.

**Comment by JingkangZhang** [comment]
@keur @jvperrin @BernardZhao 
I'll try using Vue as a template engine to remove the 1000+ lines of duplicate code. 

@BernardZhao 
For duplicate IDs that do cause problems, I changed them to classes. Otherwise if you are concerned with good HTML styles, I may just change them to classes as well, or better remove them since these adobe-generated, unreferenced IDs are unnecessary in the first place. I don't see the need to do that for now. 

Also I'll fix my Vscode format-on-save and revert the style changes so the code would be easier to review.
Thanks for y'all's comments!

**Comment by JingkangZhang** [comment]
Please see #18

**Comment by dkess** [comment]
You should *not* use Vue for this-- Vue is for interactive applications, which this is not. There are better ways to cut down on repetition than pulling in a massive frontend framework.

**Comment by JingkangZhang** [comment]
> You should _not_ use Vue for this-- Vue is for interactive applications, which this is not. There are better ways to cut down on repetition than pulling in a massive frontend framework.

What about https://github.com/janl/mustache.js?

**Comment by dkess** [comment]
While I don't fully understand how you are trying to do this, I am also still not convinced it's necessary to pull in any Javascript libraries for this. Can't you get the root element object of the scene, copy it, and then put it inside the TV?

**Comment by JingkangZhang** [comment]
> While I don't fully understand how you are trying to do this, I am also still not convinced it's necessary to pull in any Javascript libraries for this. Can't you get the root element object of the scene, copy it, and then put it inside the TV?

Code inside TV differs in 30+ places with the root SVG.

**Comment by dkess** [comment]
Can you please explain what you mean by this?

**Comment by JingkangZhang** [comment]
> Can you please explain what you mean by this?

The final code we want is 
`<g id=“1”><g id=“2”><g><g></g></g></g></g>`
However, inserting 
`<g id=“1”><g id=“2”></g></g>`
Into itself would result in
`<g id=“1”><g id=“2”><g id=“1”><g id=“2”></g></g></g></g>`

Thus, we need a template string
`<g{{}}><g{{}}></g></g>`
That either formats to
`<g id=“1”><g id=“2”></g></g>`
Or
`<g><g></g></g>`

**Comment by dkess** [comment]
There are a lot of ways you could do that in raw Javascript. The most straightforward way would be something like this:

```js
var a = document.getElementById('1');
var b = a.cloneNode(true); // true means do a deep clone
Array.from(b.getElementsByTagName('*')).forEach((el) => {
    el.removeAttribute('id');
});

document.getElementById('2').appendChild(b);
```
You could probably do a more efficient thing by getting rid of ids altogether and using classes instead, but if we don't want to mess with the original SVG, the above should work.

**Comment by JingkangZhang** [comment]
> There are a lot of ways you could do that in raw Javascript. The most straightforward way would be something like this:
> 
> ```js
> var a = document.getElementById('1');
> var b = a.cloneNode(true); // true means do a deep clone
> Array.from(b.getElementsByTagName('*')).forEach((el) => {
>     el.removeAttribute('id');
> });
> 
> document.getElementById('2').appendChild(b);
> ```
> 
> You could probably do a more efficient thing by getting rid of ids altogether and using classes instead, but if we don't want to mess with the original SVG, the above should work.

That modification is what we felt to be a bit hacky. If not, I'd happily implement it that way. May I?

**Comment by dkess** [comment]
Personally, I wouldn't consider this any hackier than the templating stuff you were doing with Vue. This approach has the advantage of not needing to change the raw text of the outputted SVG, which could be useful if we ever "re-compile" it from the Adobe Illustrator export function. You have my permission to use it :P .

Thank you for working on this!

**Review by kkuehlz — APPROVED** [review]
Excellent. Please squash these commits before merging.

**Pull request merged**